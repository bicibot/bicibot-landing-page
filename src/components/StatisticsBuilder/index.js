import React from "react";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Counter from "../../components/Counter"
import {
    BarChart, Bar, XAxis, YAxis, CartesianGrid, Tooltip, Legend, ResponsiveContainer
} from 'recharts';
import { ResponsiveCalendar } from '@nivo/calendar'
import { ResponsiveBar } from '@nivo/bar'


const StatisticsBuilder = (props) => {

    const vehiData = [];

    const vehicles = new Set(props.data.map(report => report.invasion_vehicle))

    const calendarData = props.data.reduce((days, r) => {
      let createdAt = new Date(r.createdAt).toISOString().split('T')[0];
      if (!days.some(d => d.day === createdAt)) {
        days.push({ day: createdAt, value: 0 })
      }

      let dayToBeUpdated = days.filter(d => d.day === createdAt);
      dayToBeUpdated[0].value += 1;

      return days;
    }, [])

    const timeData = props.data.reduce((hours, r) => {
      let reportHour = new Date(r.createdAt).getHours().toString() + 'h';
      if (!hours.some(h => h.hour === reportHour)) {
        hours.push({ hour: reportHour, 'denúncias': 0 })
      }

      let hourToBeUpdated = hours.filter(h => h.hour === reportHour);
      hourToBeUpdated[0]['denúncias'] += 1;

      return hours;
    }, [])

    timeData.sort((h, h2) => h.hour.split('h')[0] - h2.hour.split('h')[0])

    vehicles.forEach(v => {
      if (v !== undefined) {
        vehiData.push({ name: v, quantity: props.data.filter(r => r.invasion_vehicle === v).length })
      }
    })

    return (
        <Row className="justify-content-md-center">
            <Col lg="2" style={{ justifyContent: "center", display: "flex" }}>
                <Counter total={props.data.length} />
            </Col>
            <Col lg="6" style={{ width: '100%', height: 300 }}>
                <ResponsiveContainer>
                    <BarChart
                        data={vehiData}
                        margin={{
                            top: 20, right: 30, left: 20, bottom: 5,
                        }}
                        layout="vertical"
                    >
                        <CartesianGrid strokeDasharray="3 3" />
                        <XAxis type="number" />
                        <YAxis type="category" dataKey="name" />
                        <Tooltip />
                        <Legend />
                        <Bar name="Tipo de veículo invasores" dataKey="quantity" fill="#5A31F8" />
                    </BarChart>
                </ResponsiveContainer>
            </Col>
            <Col lg="6" className="mr-10" style={{ height: "400px", width: "100%" }}>
                <ResponsiveCalendar
                    data={calendarData}
                    from="2019-09-24"
                    to="2020-01-20"
                    emptyColor="#eeeeee"
                    colors={['#61cdbb', '#97e3d5', '#e8c1a0', '#f47560']}
                    margin={{ top: 40, right: 40, bottom: 40, left: 40 }}
                    minValue="auto"
                    yearSpacing={40}
                    monthBorderColor="#ffffff"
                    dayBorderWidth={2}
                    dayBorderColor="#ffffff"
                    legends={[
                        {
                            anchor: 'bottom-left',
                            direction: 'row',
                            translateY: 36,
                            itemCount: 4,
                            itemWidth: 42,
                            itemHeight: 36,
                            itemsSpacing: 14,
                            itemDirection: 'right-to-left'
                        }
                    ]} />
            </Col>
            <Col lg="6" className="mr-10" style={{ height: "400px", width: "100%" }}>
                <ResponsiveBar
                    data={timeData}
                    keys={['denúncias']}
                    indexBy="hour"
                    margin={{ top: 50, right: 130, bottom: 50, left: 60 }}
                    padding={0.3}
                    axisTop={null}
                    axisRight={null}
                    axisBottom={{
                        tickSize: 5,
                        tickPadding: 5,
                        tickRotation: 0,
                        legend: 'Faixa de Horário',
                        legendPosition: 'middle',
                        legendOffset: 32
                    }}
                    axisLeft={{
                        tickSize: 5,
                        tickPadding: 5,
                        tickRotation: 0,
                        legend: 'Número de Denúncias',
                        legendPosition: 'middle',
                        legendOffset: -40
                    }}
                    labelSkipWidth={12}
                    labelSkipHeight={12}
                    legends={[
                        {
                            dataFrom: 'keys',
                            anchor: 'bottom-right',
                            direction: 'column',
                            justify: false,
                            translateX: 120,
                            translateY: 0,
                            itemsSpacing: 2,
                            itemWidth: 100,
                            itemHeight: 20,
                            itemDirection: 'left-to-right',
                            itemOpacity: 0.85,
                            symbolSize: 20,
                            effects: [
                                {
                                    on: 'hover',
                                    style: {
                                        itemOpacity: 1
                                    }
                                }
                            ]
                        }
                    ]}
                    animate={true}
                    motionStiffness={90}
                    motionDamping={15}
                />
            </Col>
        </Row>
    )
}

export default StatisticsBuilder