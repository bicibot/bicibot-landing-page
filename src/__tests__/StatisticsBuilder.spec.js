import React from "react";
import { create } from "react-test-renderer";
import StatisticsBuilder from "../components/StatisticsBuilder"

let data = [
    {
        "_id": "5d88bda107bad80023d00c21",
        "report_type": "Ameaça",
        "location": [
            -34.888514,
            -8.0444708
        ],
        "city": "Recife",
        "description": "Passou perto/rápido",
        "invasion_vehicle": "Ônibus",
        "invasion_time": "06h01 ao 12h00",
        "invasion_state": "Veiculo trafegando na ciclofaixa",
        "address": "avenida agamenon magalhães, 17, santo amaro",
        "createdAt": "2019-09-23T12:42:09.455Z",
        "updatedAt": "2019-09-23T12:42:09.455Z",
        "__v": 0
    }
]

describe("Counter component", () => {
    test("Matches the snapshot", () => {
        const statisticsBuilder = create(<StatisticsBuilder data={data} />);
        expect(statisticsBuilder.toJSON()).toMatchSnapshot();
    });
});