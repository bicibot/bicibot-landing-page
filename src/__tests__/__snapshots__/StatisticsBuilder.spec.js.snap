// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component Matches the snapshot 1`] = `
<div
  className="justify-content-md-center row"
  style={
    Object {
      "marginTop": "50px",
    }
  }
>
  <div
    className="col-lg-2"
    style={
      Object {
        "display": "flex",
        "justifyContent": "center",
        "textAlign": "center",
      }
    }
  >
    <div
      className="card"
      style={
        Object {
          "height": "fit-content",
        }
      }
    >
      <div
        className="sc-htpNat dFcVJV"
      >
        <div
          className="sc-bwzfXH fpBhud"
        >
          <div
            className="sc-bdVaJa qkxPG"
          >
            1
          </div>
        </div>
        <h2>
          den√∫ncias
        </h2>
      </div>
    </div>
  </div>
  <div
    className="col-lg-6"
    style={
      Object {
        "height": 300,
        "width": "100%",
      }
    }
  >
    <div
      className="recharts-responsive-container"
      id="undefined"
      style={
        Object {
          "height": "100%",
          "maxHeight": undefined,
          "minHeight": undefined,
          "minWidth": undefined,
          "width": "100%",
        }
      }
    >
      <div />
    </div>
  </div>
  <div
    className="mr-10 col-lg-6"
    style={
      Object {
        "height": "400px",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </div>
  <div
    className="mr-10 col-lg-6"
    style={
      Object {
        "height": "400px",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </div>
</div>
`;
